# **Eda**

## 1. Data agregation

In [1]:
import pandas as pd
from IPython.display import display
import os

In [2]:
DATA_NAMES = os.listdir('./datasets')

In [3]:
print(f'Total: {len(DATA_NAMES)}') 
print(f'Data names: {DATA_NAMES}') 

Total: 9
Data names: ['D_loan.csv', 'D_clients.csv', 'D_last_credit.csv', 'D_pens.csv', 'D_salary.csv', 'D_target.csv', 'D_work.csv', 'D_job.csv', 'D_close_loan.csv']


In [4]:
df1, df2, df3, df4, df5, df6, df7, df8, df9 = [pd.read_csv('datasets/'+i) for i in DATA_NAMES]

data_list = [df1, df2, df3, df4, df5, df6, df7, df8, df9]

In [5]:
for data_number in range(len(data_list)):
    print(f'DATA NUMBER {data_number}')
    display(data_list[data_number].head())
    print('Short data description:')
    display(data_list[data_number].info())
    display(data_list[data_number].nunique())
    

DATA NUMBER 0


ID_LOAN  ID_CLIENT
0  1753790658  106804370
1  1753790659  106804371
2  1753790660  106804372
3  1753790661  106804372
4  1753790662  106804373

Short data description:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21126 entries, 0 to 21125
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype
---  ------     --------------  -----
 0   ID_LOAN    21126 non-null  int64
 1   ID_CLIENT  21126 non-null  int64
dtypes: int64(2)
memory usage: 330.2 KB


None

ID_LOAN      21126
ID_CLIENT    15223
dtype: int64

DATA NUMBER 1


ID  AGE  GENDER            EDUCATION      MARITAL_STATUS  \
0  106805103   42       1              Среднее  Не состоял в браке   
1  106809308   28       1  Среднее специальное      Состою в браке   
2  106805867   64       0  Среднее специальное      Состою в браке   
3  106808779   54       1  Среднее специальное      Состою в браке   
4  106814289   26       0  Среднее специальное      Состою в браке   

   CHILD_TOTAL  DEPENDANTS  SOCSTATUS_WORK_FL  SOCSTATUS_PENS_FL  \
0            1           0                  1                  0   
1            1           1                  1                  0   
2            2           0                  1                  1   
3            0           0                  1                  0   
4            1           1                  1                  0   

    REG_ADDRESS_PROVINCE  FACT_ADDRESS_PROVINCE POSTAL_ADDRESS_PROVINCE  \
0     Московская область     Московская область      Московская область   
1      Читинская область      Читинская область       Читинская область   
2      Иркутская область      Иркутская область       Иркутская область   
3  Новосибирская область  Новосибирская область   Новосибирская область   
4      Красноярский край      Красноярский край       Красноярский край   

   FL_PRESENCE_FL  OWN_AUTO  
0               1         0  
1               0         0  
2               0         1  
3               1         1  
4               1         0

Short data description:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16000 entries, 0 to 15999
Data columns (total 14 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   ID                       16000 non-null  int64 
 1   AGE                      16000 non-null  int64 
 2   GENDER                   16000 non-null  int64 
 3   EDUCATION                16000 non-null  object
 4   MARITAL_STATUS           16000 non-null  object
 5   CHILD_TOTAL              16000 non-null  int64 
 6   DEPENDANTS               16000 non-null  int64 
 7   SOCSTATUS_WORK_FL        16000 non-null  int64 
 8   SOCSTATUS_PENS_FL        16000 non-null  int64 
 9   REG_ADDRESS_PROVINCE     16000 non-null  object
 10  FACT_ADDRESS_PROVINCE    16000 non-null  object
 11  POSTAL_ADDRESS_PROVINCE  16000 non-null  object
 12  FL_PRESENCE_FL           16000 non-null  int64 
 13  OWN_AUTO                 16000 non-null  int64 
dtypes: int64(9), o

None

ID                         16000
AGE                           47
GENDER                         2
EDUCATION                      7
MARITAL_STATUS                 5
CHILD_TOTAL                   10
DEPENDANTS                     8
SOCSTATUS_WORK_FL              2
SOCSTATUS_PENS_FL              2
REG_ADDRESS_PROVINCE          81
FACT_ADDRESS_PROVINCE         81
POSTAL_ADDRESS_PROVINCE       80
FL_PRESENCE_FL                 2
OWN_AUTO                       3
dtype: int64

DATA NUMBER 2


CREDIT  TERM  FST_PAYMENT  ID_CLIENT
0   8000.00     6       8650.0  106804370
1  21650.00     6       4000.0  106804371
2  33126.00    12       4000.0  106804372
3   8491.82     6       5000.0  106804373
4  21990.00    12       4000.0  106804374

Short data description:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15223 entries, 0 to 15222
Data columns (total 4 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   CREDIT       15223 non-null  float64
 1   TERM         15223 non-null  int64  
 2   FST_PAYMENT  15223 non-null  float64
 3   ID_CLIENT    15223 non-null  int64  
dtypes: float64(2), int64(2)
memory usage: 475.8 KB


None

CREDIT          8029
TERM              22
FST_PAYMENT     2937
ID_CLIENT      15223
dtype: int64

DATA NUMBER 3


ID  FLAG       COMMENT
0   1     0     пенсионер
1   2     1  не пенсионер

Short data description:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2 entries, 0 to 1
Data columns (total 3 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   ID       2 non-null      int64 
 1   FLAG     2 non-null      int64 
 2   COMMENT  2 non-null      object
dtypes: int64(2), object(1)
memory usage: 176.0+ bytes


None

ID         2
FLAG       2
COMMENT    2
dtype: int64

DATA NUMBER 4


FAMILY_INCOME  PERSONAL_INCOME  ID_CLIENT
0  от 20000 до 50000 руб.          20000.0  106809321
1  от 20000 до 50000 руб.          14000.0  106815561
2  от 10000 до 20000 руб.          15000.0  106811521
3  от 20000 до 50000 руб.          20000.0  106811252
4  от 20000 до 50000 руб.          25000.0  106808620

Short data description:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15523 entries, 0 to 15522
Data columns (total 3 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   FAMILY_INCOME    15523 non-null  object 
 1   PERSONAL_INCOME  15523 non-null  float64
 2   ID_CLIENT        15523 non-null  int64  
dtypes: float64(1), int64(1), object(1)
memory usage: 363.9+ KB


None

FAMILY_INCOME          5
PERSONAL_INCOME      274
ID_CLIENT          15223
dtype: int64

DATA NUMBER 5


AGREEMENT_RK  ID_CLIENT  TARGET
0      59910150  106804370       0
1      59910230  106804371       0
2      59910525  106804372       0
3      59910803  106804373       0
4      59911781  106804374       0

Short data description:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15223 entries, 0 to 15222
Data columns (total 3 columns):
 #   Column        Non-Null Count  Dtype
---  ------        --------------  -----
 0   AGREEMENT_RK  15223 non-null  int64
 1   ID_CLIENT     15223 non-null  int64
 2   TARGET        15223 non-null  int64
dtypes: int64(3)
memory usage: 356.9 KB


None

AGREEMENT_RK    15223
ID_CLIENT       15223
TARGET              2
dtype: int64

DATA NUMBER 6


ID  FLAG      COMMENT
0   1     0     работает
1   2     1  не работает
2   3     2  не известно

Short data description:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3 entries, 0 to 2
Data columns (total 3 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   ID       3 non-null      int64 
 1   FLAG     3 non-null      int64 
 2   COMMENT  3 non-null      object
dtypes: int64(2), object(1)
memory usage: 200.0+ bytes


None

ID         3
FLAG       3
COMMENT    3
dtype: int64

DATA NUMBER 7


GEN_INDUSTRY                    GEN_TITLE  \
0                   Торговля                      Рабочий   
1                   Торговля                      Рабочий   
2  Информационные технологии                   Специалист   
3                Образование  Руководитель среднего звена   
4     Государственная служба                   Специалист   

                         JOB_DIR  WORK_TIME  ID_CLIENT  
0    Вспомогательный техперсонал       18.0  106804370  
1  Участие в основ. деятельности       97.0  106804371  
2  Участие в основ. деятельности       84.0  106804372  
3  Участие в основ. деятельности      168.0  106804373  
4  Участие в основ. деятельности      101.0  106804374

Short data description:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15223 entries, 0 to 15222
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   GEN_INDUSTRY  13856 non-null  object 
 1   GEN_TITLE     13856 non-null  object 
 2   JOB_DIR       13856 non-null  object 
 3   WORK_TIME     13855 non-null  float64
 4   ID_CLIENT     15223 non-null  int64  
dtypes: float64(1), int64(1), object(3)
memory usage: 594.8+ KB


None

GEN_INDUSTRY       31
GEN_TITLE          12
JOB_DIR            10
WORK_TIME         302
ID_CLIENT       15223
dtype: int64

DATA NUMBER 8


ID_LOAN  CLOSED_FL
0  1753790658          1
1  1753790659          1
2  1753790660          1
3  1753790661          0
4  1753790662          1

Short data description:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21126 entries, 0 to 21125
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype
---  ------     --------------  -----
 0   ID_LOAN    21126 non-null  int64
 1   CLOSED_FL  21126 non-null  int64
dtypes: int64(2)
memory usage: 330.2 KB


None

ID_LOAN      21126
CLOSED_FL        2
dtype: int64

**The datasets contain:**

1. Information about `21126` loans from `15223` clients.
2. Information about social features of `16000` clients.
3. Information abot loan details related to `15223` clients.
4. Dataset with for encoding retirement status. (0 — retirement=True, 1 - retirement=False)
5. Dataset with family and personal income of `15223` clients.
6. Dataset with target feature to `15223` clients.
7. Dataset with for encoding work status. (0 — works=True, 1 - works=False, 2 — works=None)
8. Information about work of `15223` clients, contains NANs
9. Infromation about loan closing status. 

## 2. Data merging

- It is necessary to separate the data connection by the ID_LOAN and ID_CLIENT key
- Data with ID_CLIENT key: df1, df2, df3, df5, df6, df8
- Data with ID_LOAN key: df1, df9

Merging steps:

**First:** Merging and features extracting using ID_LOAN.

1) Left merge loan code `df1` and loan closing `df9` data, left because we can have old data about the closing of the loan -> `loans_merged_df`
2) Extracting features about total conut of loans and total count of closed loans for each client -> `loans_df`

**Second:** Merging and features extracting using ID_CLIENT.

1) Left merge loan details `df6` and target `df3`, because to train our model we need information about clients with target feature -> general_df
2)  Left merge `general_df` and social fetures `df2` -> general_df
3)  Outer merge `general_df`  and personal income `df5` data -> general_df
4)  Outer merge `general_df` and work status `df8` data -> general_df

We use left merge in steps 1-4, because we need information ONLY about clients with target feature 

**Third:**

1) Left merge `general_df` and `loans_df`,  because we need information ONLY loans clients with target feature 

In [156]:
#ID_LOAN
loans_merged_df = df1.merge(df9, how='left', on='ID_LOAN')

#EXTRAXTING FEATURES

loans_df = loans_merged_df.groupby('ID_CLIENT').agg({'CLOSED_FL' : 'sum', 'ID_LOAN' : 'count'})
loans_df.columns = ['closed_loans_count', 'total_loans_count']

loans_df.reset_index(inplace=True)

loans_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15223 entries, 0 to 15222
Data columns (total 3 columns):
 #   Column              Non-Null Count  Dtype
---  ------              --------------  -----
 0   ID_CLIENT           15223 non-null  int64
 1   closed_loans_count  15223 non-null  int64
 2   total_loans_count   15223 non-null  int64
dtypes: int64(3)
memory usage: 356.9 KB


In [157]:
#ID_CLIENT

#Correction of the key in the data about social features
df2_columns_list = list(df2.columns)
df2_columns_list[0] = 'ID_CLIENT'

df2.columns = df2_columns_list

#Data merging

general_df = df6.merge(df3, how='left', on='ID_CLIENT') # joined target and loan details data

general_df = general_df.merge(df2, how='left', on='ID_CLIENT') # joined social features and loan details

general_df = general_df.merge(df5, how='left', on='ID_CLIENT') # joined general_df and income details
general_df = general_df.merge(df8, how='left', on='ID_CLIENT')# joined work status data

general_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15523 entries, 0 to 15522
Data columns (total 25 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   AGREEMENT_RK             15523 non-null  int64  
 1   ID_CLIENT                15523 non-null  int64  
 2   TARGET                   15523 non-null  int64  
 3   CREDIT                   15523 non-null  float64
 4   TERM                     15523 non-null  int64  
 5   FST_PAYMENT              15523 non-null  float64
 6   AGE                      15523 non-null  int64  
 7   GENDER                   15523 non-null  int64  
 8   EDUCATION                15523 non-null  object 
 9   MARITAL_STATUS           15523 non-null  object 
 10  CHILD_TOTAL              15523 non-null  int64  
 11  DEPENDANTS               15523 non-null  int64  
 12  SOCSTATUS_WORK_FL        15523 non-null  int64  
 13  SOCSTATUS_PENS_FL        15523 non-null  int64  
 14  REG_ADDRESS_PROVINCE  

In [158]:
#LOANS AND CLIENTS DATA MERGING

general_df = general_df.merge(loans_df, on='ID_CLIENT', how='left')

general_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15523 entries, 0 to 15522
Data columns (total 27 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   AGREEMENT_RK             15523 non-null  int64  
 1   ID_CLIENT                15523 non-null  int64  
 2   TARGET                   15523 non-null  int64  
 3   CREDIT                   15523 non-null  float64
 4   TERM                     15523 non-null  int64  
 5   FST_PAYMENT              15523 non-null  float64
 6   AGE                      15523 non-null  int64  
 7   GENDER                   15523 non-null  int64  
 8   EDUCATION                15523 non-null  object 
 9   MARITAL_STATUS           15523 non-null  object 
 10  CHILD_TOTAL              15523 non-null  int64  
 11  DEPENDANTS               15523 non-null  int64  
 12  SOCSTATUS_WORK_FL        15523 non-null  int64  
 13  SOCSTATUS_PENS_FL        15523 non-null  int64  
 14  REG_ADDRESS_PROVINCE  

Data Merging Conclusions:

1) 